### PR TITLE
linuxPackages.nvidia_x11.libXNVCtrl: make the shared library available

### PR DESCRIPTION
**Pull request by aidalgol**
## Description of changes
The libXNVCtrl library is used by several programs in nixpkgs to get Nvidia GPU stats, and they all load the library dynamically, but the libXNVCtrl package only builds a static library.  There is no build flag to enable this, so we have to patch the Makefile to produce a shared library.  We then add the package to `hardware.opengl.extraPackages` in the NixOS module `hardware/video/nvidia.nix` so that it is available on Nvidia-enabled systems.

Based on a patch by @negativo17 for Fedora: https://github.com/negativo17/nvidia-settings/blob/master/nvidia-settings-libXNVCtrl.patch

## Things done
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/os-specific/linux/nvidia-x11/settings.nix
Comment (ConnorBaker) [inline pkgs/os-specific/linux/nvidia-x11/settings.nix:52]: Function application has a higher precedence than equality so there's no need for parenthesis around the calls to `lib.versions.major`.

**Review by ConnorBaker — COMMENTED** [review]
One minor nit; I can try the builds later to verify they build.

Do you have an easy way to verify this is working as intended?

Also, would you mind adding yourself to the list of maintainers for pkgs/os-specific/linux/nvidia-x11/settings.nix?

File: pkgs/os-specific/linux/nvidia-x11/libxnvctrl-build-shared-340.patch
Comment (SomeoneSerge) [inline pkgs/os-specific/linux/nvidia-x11/libxnvctrl-build-shared-340.patch:1]: Looking at https://github.com/NVIDIA/nvidia-settings/blob/4e4bb3fecbbdd902d6f4bc6bbfb9902043c5c7f6/src/libXNVCtrl/Makefile#L39-L53, it seems that we could build the `.so` just by invoking `make` with appropriate arguments. I think that'd be more maintainable than patching `Makefile` per-release.

If we wish to build both the static and the shared object in a single derivation we'd still have to define a custom `buildPhase`, but I think that's OK.

**Comment by aidalgol** [comment]
> Do you have an easy way to verify this is working as intended?

The program I was trying to get to work with my nvidia GPU was MangoHUD, so running `vkcube` with that was how I was testing my fix.

Before changes:
```console
❯ mangohud vkcube
Selected GPU 0: NVIDIA GeForce RTX 3080, type: DiscreteGpu
[2024-05-20 06:44:07.254] [MANGOHUD] [error] [loader_nvctrl.cpp:39] Failed to open 64bit libXNVCtrl.so.0: libXNVCtrl.so.0: cannot open shared object file: No such file or directory
[2024-05-20 06:44:07.254] [MANGOHUD] [error] [nvctrl.cpp:48] XNVCtrl loader failed to load
```

After changes:
```console
❯ mangohud vkcube
Selected GPU 0: NVIDIA GeForce RTX 3080, type: DiscreteGpu
```

(With `gpu_stats` in my `~/.config/MangoHud/MangoHud.conf`.)

File: pkgs/os-specific/linux/nvidia-x11/libxnvctrl-build-shared-340.patch
Comment (aidalgol) [reply pkgs/os-specific/linux/nvidia-x11/libxnvctrl-build-shared-340.patch:1]: There's no make rule for a `.so` file in `xnvctrl.mk` (or anywhere that I can see), so I don't think we can do this via `make` arguments.  We could possibly just bypass `make` entirely and invoke the commands directly in a custom `buildPhase`, but I'm not sure whether that would be more maintainable.

File: pkgs/os-specific/linux/nvidia-x11/settings.nix
Comment (aidalgol) [reply pkgs/os-specific/linux/nvidia-x11/settings.nix]: Fixed.

File: pkgs/os-specific/linux/nvidia-x11/libxnvctrl-build-shared-340.patch
Comment (SomeoneSerge) [reply pkgs/os-specific/linux/nvidia-x11/libxnvctrl-build-shared-340.patch:1]: Oh I notice the `$(AR)` part, I was naive to think they had some kind of a generic rule for linkage...

Comment (SomeoneSerge) [reply pkgs/os-specific/linux/nvidia-x11/libxnvctrl-build-shared-340.patch:1]: I'd still like to get rid of the conditional patches. A part of the diff doesn't seem to be essential to the derivation, e.g. the `clean` and `all` rules. Maybe omitting those we could have a single patch that successfully applies to all releases: adding nothing but the `.so` rule.

Comment (aidalgol) [reply pkgs/os-specific/linux/nvidia-x11/libxnvctrl-build-shared-340.patch:1]: I was able to simplify down to two patches, because `xnvctrl.mk` does not exist in the `3xx` versions.

**Review by SomeoneSerge — APPROVED** [review]
I've some general comments left, in the line of "are we solving the right problem":

1. It seems that patching `Makefile`s takes approximately the same amount of code as shipping our own ad hoc `CMakeLists.txt` would have...
2. It seems that upstream does not distribute shared libraries, so it feels kind of arbitrary that MangoHUD should use it as one.
3. Further on MangoHUD, these probably should be in `buildInputs`: https://github.com/NixOS/nixpkgs/blob/dbeb624729a7c6308d6a945b0a1fbafd001c2813/pkgs/tools/graphics/mangohud/default.nix#L190-L194

That said, if you @aidalgol are convinced this is the way to address your use-case, I'll merge

**Comment by aidalgol** [comment]
> 1. It seems that patching `Makefile`s takes approximately the same amount of code as shipping our own ad hoc `CMakeLists.txt` would have...

Do you mean for all of `nvidia_x11`, or just `libXNVCtrl`?

> 2. It seems that upstream does not distribute shared libraries, so it feels kind of arbitrary that MangoHUD should use it as one.

MangoHUD is not the only package in nixpkgs that references `libXNVCtrl`.

- `psensor`
- `nvfancontrol`
- `monitor` (yes, it's really called just "monitor")
- `vibrantlinux`
- `conky`

But I don't know whether any of those try to link dynamically.  It sounds as if Conky does: [Installing libxnvctrl (525.60.11-2) breaks dependency ‘libXNVCtrl.so=0-64’ required by conky](https://forum.manjaro.org/t/installing-libxnvctrl-525-60-11-2-breaks-dependency-libxnvctrl-so-0-64-required-by-conky/127981/2).

Looking at other Linux distributions, at least Arch patches `libXNVCtrl` similarly.

- https://bugs.archlinux.org/task/76702
- https://archlinux.org/packages/extra/x86_64/libxnvctrl/

> 3. Further on MangoHUD, these probably should be in `buildInputs`: https://github.com/NixOS/nixpkgs/blob/dbeb624729a7c6308d6a945b0a1fbafd001c2813/pkgs/tools/graphics/mangohud/default.nix#L190-L194

That I'm not too sure about.  If someone is using `mangohud` installed via "standalone" home-manager, and they update that more frequently than their system configuration, wouldn't `mangohud` have a newer version than the system's driver?

**Comment by SomeoneSerge** [comment]
> Do you mean for all of nvidia_x11, or just libXNVCtrl?

libXNVCtrl

> Looking at other Linux distributions, at least Arch patches libXNVCtrl similarly.

Seems like the patch belongs in NVIDA/nvidia-settings then, let's aim to eventually move it there.

**Comment by SomeoneSerge** [comment]
> You mean upstream it?

Yes

> In that case, I would not be opposed to instead replacing the upstream Makefile, if you think that would be simpler or easier to maintain.

I'd be in favour of that. But also if we're upstreaming a patch, we'll want to switch to `fetchpatch`, and it's safer to avoid trying to impose cmake (or anything) on people...

If you need time to choose but you're committed to following up on this, I can merge the current PR as a temporary fix

**Comment by aidalgol** [comment]
I have submitted a PR upstream and have changed the `libXNVCtrl` derivation to use the diff from the PR, except for the `3xx` legacy version.

File: pkgs/os-specific/linux/nvidia-x11/settings.nix
Comment (SomeoneSerge) [inline pkgs/os-specific/linux/nvidia-x11/settings.nix:52]: - Iirc the `pull/xxxx/commits/yyyy` URL may get invalidated after force-pushing or even rebase-merging (I may be wrong)
- The `fetchpatch` now has an updated `all` rule, but the 3xx patch doesn't

Comment (aidalgol) [reply pkgs/os-specific/linux/nvidia-x11/settings.nix:52]: I think you're right.  I thought that would make it robust against PR changes, but I think GitHub does invalidate commit URLs on a force push, at least after a while.  I think I should back out this commit until the PR is merged.

File: pkgs/os-specific/linux/nvidia-x11/settings.nix
Comment (SomeoneSerge) [inline pkgs/os-specific/linux/nvidia-x11/settings.nix:47]: Since there's a merge conflict anyway, another nit: `makeTargets` is more explicit

Comment (aidalgol) [reply pkgs/os-specific/linux/nvidia-x11/settings.nix:47]: I tried `makeTargets` first, and it did not work.  I don't see any mention of this attribute in the [nixpkgs manual](https://nixos.org/manual/nixpkgs/unstable/#var-stdenv-makeFlags).

Comment (SomeoneSerge) [reply pkgs/os-specific/linux/nvidia-x11/settings.nix:47]: Hm, I was sure I saw the attribute being used somewhere, but I don't see any mentions in `pkgs/stdenv/generic/setup.sh` so I must be wrong

File: nixos/modules/hardware/video/nvidia.nix
Comment (SomeoneSerge) [inline nixos/modules/hardware/video/nvidia.nix:1]: Let's solve conflicts, merge, and backport to the last two releases (counting 24.05)

Comment (aidalgol) [reply nixos/modules/hardware/video/nvidia.nix:1]: Branch rebased on recent `master`.

File: nixos/modules/hardware/video/nvidia.nix
Comment (SomeoneSerge) [inline nixos/modules/hardware/video/nvidia.nix:288]: I had another look at the derivation and it actually doesn't seem to depend on `nvidia_x11`, except through `nvidia_x11.settingsVersion`. This makes me question if we actually have to deploy this impurely rather than link directly

Comment (megheaiulian) [inline nixos/modules/hardware/video/nvidia.nix:288]: :exclamation:  This will break things. Depending on config ( `useSettings` or `nvidiaSettings` ) `nvidia_x11.settings` might be a empty set.  See https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/os-specific/linux/nvidia-x11/generic.nix#L215
This should handle that case or be reverted.

Comment (aidalgol) [reply nixos/modules/hardware/video/nvidia.nix:288]: Submitted https://github.com/NixOS/nixpkgs/pull/316516 to address this.

Comment (SomeoneSerge) [reply nixos/modules/hardware/video/nvidia.nix:288]: @megheaiulian thanks for the heads up!

**Pull request merged**
